security/advancedtls/advancedtls_test.go

```diff
@@ -46,14 +47,249 @@ func Test(t *testing.T) {
+	if err != nil {
+		return nil, fmt.Errorf("failed to load certs: %v", err)
+	}
+	if fake.isClient {
+		return &certprovider.KeyMaterial{Roots: cs.clientTrust1}, nil
+	}
+	return &certprovider.KeyMaterial{Roots: cs.serverTrust1}, nil
+}
+
+func (fake fakeRootProvider) Close() {}
+
+type fakeIdentityProvider struct {
+	isClient      bool
+	wantMultiCert bool
+	wantError     bool
+}
+
+func (fake fakeIdentityProvider) KeyMaterial(ctx context.Context) (*certprovider.KeyMaterial, error) {
+	cs := &certStore{}
+	err := cs.loadCerts()
+	if fake.wantError {
+		return nil, fmt.Errorf("bad fakeIdentityProvider")
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to load certs: %v", err)
+	}
+	if fake.isClient {
+		if fake.wantMultiCert {
+			return &certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1, cs.clientPeer2}}, nil
+		}
+		return &certprovider.KeyMaterial{Certs: []tls.Certificate{cs.clientPeer1}}, nil
+	}
+	if fake.wantMultiCert {
+		return &certprovider.KeyMaterial{Certs: []tls.Certificate{cs.serverPeer1, cs.serverPeer2}}, nil
+	}
+	return &certprovider.KeyMaterial{Certs: []tls.Certificate{cs.serverPeer1}}, nil
+}
+
+func (fake fakeIdentityProvider) Close() {}
+
+func (s) TestClientOptionsConfig(t *testing.T) {
+	tests := []struct {
+		desc        string
+		clientVType VerificationType
+		IdentityCertificateOptions
+		RootCertificateOptions
+		wantError bool
+	}{
+		{
+			desc:        "Skip default verification and provide no root credentials",
+			clientVType: SkipVerification,
+			wantError:   true,
+		},
+		{
+			desc:        "More than one fields in RootCertificateOptions is specified",
+			clientVType: CertVerification,
+			RootCertificateOptions: RootCertificateOptions{
+				RootCACerts:  x509.NewCertPool(),
+				RootProvider: fakeRootProvider{},
+			},
+			wantError: true,
+		},
+		{
+			desc:        "More than one fields in IdentityCertificateOptions is specified",
+			clientVType: CertVerification,
+			IdentityCertificateOptions: IdentityCertificateOptions{
+				GetIdentityCertificatesForClient: func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+					return nil, nil
+				},
+				IdentityProvider: fakeIdentityProvider{},
+			},
+			wantError: true,
+		},
+		{
+			desc: "Specify GetIdentityCertificatesForServer",
+			IdentityCertificateOptions: IdentityCertificateOptions{
+				GetIdentityCertificatesForServer: func(*tls.ClientHelloInfo) ([]*tls.Certificate, error) {
+					return nil, nil
+				},
+			},
+			wantError: true,
+		},
+		{
+			desc:        "Use system default if no fields in RootCertificateOptions is specified",
+			clientVType: CertVerification,
+		},
+		{
+			desc:        "Good case with mutual TLS",
+			clientVType: CertVerification,
+			RootCertificateOptions: RootCertificateOptions{
+				RootProvider: fakeRootProvider{},
+			},
+			IdentityCertificateOptions: IdentityCertificateOptions{
+				IdentityProvider: fakeIdentityProvider{},
+			},
+		},
+	}
+	for _, test := range tests {
```

Comment:
You could also consider breaking this test into two, one for the success cases and one for the failure cases. That will simplify the actual test code inside `t.Run`.